unit_test/sparse/Test_Sparse_spmv.hpp

```diff
@@ -93,15 +93,24 @@ void sequential_spmv(crsMat_t input_mat, x_vector_type x, y_vector_type y,
 
 template <typename crsMat_t, typename x_vector_type, typename y_vector_type>
 void check_spmv(crsMat_t input_mat, x_vector_type x, y_vector_type y,
-    typename y_vector_type::non_const_value_type alpha, typename y_vector_type::non_const_value_type beta){
+                typename y_vector_type::non_const_value_type alpha,
+                typename y_vector_type::non_const_value_type beta){
   //typedef typename crsMat_t::StaticCrsGraphType graph_t;
   typedef typename crsMat_t::execution_space ExecSpace;
   typedef Kokkos::RangePolicy<ExecSpace> my_exec_space;
 
-  typedef typename crsMat_t::values_type::non_const_type scalar_view_t;
-  typedef typename scalar_view_t::value_type ScalarA;
-  double eps = std::is_same<ScalarA,float>::value?2*1e-3:1e-7;
-  size_t nr = input_mat.numRows();
+  // y is the quantity being tested here,
+  // so let us use y_value_type to determine
+  // the appropriate tolerance precision.
+  using y_value_type  = typename y_vector_type::non_const_value_type;
+  using y_value_trait = Kokkos::ArithTraits<y_value_type>;
+  // Note on 2019-12-03 by LBV:
+  // one really want the following definition,
+  // const y_value_trait::mag_type eps = f(y_value_trait::eps());
+  // where f() introduces a fudge factor that accounts for
+  // rounding error and such...
+  const double eps = std::is_same<y_value_trait::mag_type, float>::value ? 2*1e-3 : 1e-7;
```

Comment:
Wow, both of those tolerances are pretty loose.  I know this was there before, but should it really be this loose?

Reply:
agreed, as you can see in my comment I am not fond of what is currently implemented.
I feel something like `100*y_value_trait::eps()` should be working, @ndellingwood any chance you could try with that?

Reply:
@lucbv we tested with 
`1000*y_value_trait::eps()` and spmv_struct tests passed, seeing lots of failures with spmv after these changes:

```
[  FAILED  ] cuda.sparse_spmv_double_int_int_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_double_int_size_t_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_float_int_int_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_float_int_size_t_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_kokkos_complex_double_int_int_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_kokkos_complex_double_int_size_t_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_kokkos_complex_float_int_int_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_kokkos_complex_float_int_size_t_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_mv_double_int_int_LayoutLeft_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_mv_double_int_size_t_LayoutLeft_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_mv_float_int_int_LayoutLeft_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_mv_float_int_size_t_LayoutLeft_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_mv_kokkos_complex_double_int_int_LayoutLeft_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_mv_kokkos_complex_double_int_size_t_LayoutLeft_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_mv_kokkos_complex_float_int_int_LayoutLeft_TestExecSpace
[  FAILED  ] cuda.sparse_spmv_mv_kokkos_complex_float_int_size_t_LayoutLeft_TestExecSpace
```

Reply:
Are these large problems, with a lot of rows or entries per row?

Reply:
I am actually considering doing something like this:
```
const double eps = max_nnz_per_row*y_vector_trait::eps()
```
This would mean that we tolerate an accumulation of round-off error which is equivalent to the number of additions happening to compute a single entry in `y`, does that make sense?

Reply:
@lucbv makes sense for infinity-norm error, which I think is the right metric here.